File: lecture09/exercise09-B/template.py
#!/usr/bin/env python3

# Exercise 09-B: Binary Tree Min

from dataclasses import dataclass
from typing import Optional
import sys

# Structures

@dataclass
class Node:
    value: int
    left:  Optional['Node']
    right: Optional['Node']

# Functions

def minimum_tree(root: Optional[Node]) -> int:
    ''' Use divide and conquer to compute the minimum of binary tree '''
    # Base case: Invalid Node

    # Divide and Conquer: Recursively solve left and right sub-trees

    # Combine: Take minimum of current node and left and right minimums
    return 0

# Main Execution

def main() -> None:
    # Create tree
    root = Node(7,
        Node(5,
                Node(3, None, None),
                None),
        Node(4,
                None,
                Node(2, None, None),
    ))

    # Compute minimum of tree
    print(minimum_tree(root))

if __name__ == '__main__':
    main()

File: lecture09/exercise09-B/template.cpp
// Exercise 09-B: Binary Tree Min

#include <climits>
#include <iostream>
#include <memory>

using namespace std;

// Node structures

template <typename T>
struct Node {
    T     value;
    Node *left;
    Node *right;
};

// Find minimum of tree

template <typename T>
T minimum_tree(Node<T> *root) {
    // Base case: Invalid Node

    // Divide and Conquer: Recursively solve left and right sub-trees

    // Combine: Take minimum of current node and left and right minimums
    return 0;
}

// Main execution

int main(int argc, char *argv[]) {
    Node<int> *root = new Node<int>{7,
    	new Node<int>{5,
    	    new Node<int>{3, nullptr, nullptr},
    	    nullptr
	},
    	new Node<int>{4,
    	    nullptr,
    	    new Node<int>{2, nullptr, nullptr}
	}
    };

    cout << minimum_tree(root) << endl;
    return 0;
}
